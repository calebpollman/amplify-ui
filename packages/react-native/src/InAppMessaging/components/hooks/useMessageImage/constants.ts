/*
 * Copyright 2017-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

import { Dimensions } from 'react-native';
import { ImageDimensions } from './types';

// as images are not expected to be responsive to orientation changes get screen dimensions at app start
const SCREEN_DIMENSIONS = Dimensions.get('screen');

// compare screen width and height and assign the lesser of the two as the base screen dimension
const BASE_SCREEN_DIMENSION =
  SCREEN_DIMENSIONS.width < SCREEN_DIMENSIONS.height
    ? SCREEN_DIMENSIONS.width
    : SCREEN_DIMENSIONS.height;

// base size that message images should fill
// - all banner message images should fill 20 percent of the base screen dimension
// - all other components should fill 60 percent of the base screen dimension
export const BANNER_IMAGE_SCREEN_MULTIPLIER = 0.2;
export const CAROUSEL_IMAGE_SCREEN_MULTIPLIER = 0.6;
export const FULL_SCREEN_IMAGE_SCREEN_MULTIPLIER = 0.6;
export const MODAL_IMAGE_SCREEN_MULTIPLIER = 0.6;

export const BANNER_IMAGE_SCREEN_SIZE =
  BANNER_IMAGE_SCREEN_MULTIPLIER * BASE_SCREEN_DIMENSION;
export const CAROUSEL_IMAGE_SCREEN_SIZE =
  CAROUSEL_IMAGE_SCREEN_MULTIPLIER * BASE_SCREEN_DIMENSION;
export const FULL_SCREEN_IMAGE_SCREEN_SIZE =
  FULL_SCREEN_IMAGE_SCREEN_MULTIPLIER * BASE_SCREEN_DIMENSION;
export const MODAL_IMAGE_SCREEN_SIZE =
  MODAL_IMAGE_SCREEN_MULTIPLIER * BASE_SCREEN_DIMENSION;

export const INITIAL_IMAGE_DIMENSIONS: ImageDimensions = {
  height: null,
  width: null,
};
