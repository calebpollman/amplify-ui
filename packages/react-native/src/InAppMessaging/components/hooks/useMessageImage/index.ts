export { default } from './useMessageImage';
export { ImageDimensions } from './types';
