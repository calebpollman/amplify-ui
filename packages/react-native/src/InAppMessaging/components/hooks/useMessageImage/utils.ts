/*
 * Copyright 2017-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

import { Image } from 'react-native';
import { ConsoleLogger as Logger } from '@aws-amplify/core';
import { InAppMessageLayout } from '@aws-amplify/notifications';

import {
  BANNER_IMAGE_SCREEN_SIZE,
  CAROUSEL_IMAGE_SCREEN_SIZE,
  FULL_SCREEN_IMAGE_SCREEN_SIZE,
  MODAL_IMAGE_SCREEN_SIZE,
} from './constants';
import { ImageDimensions, ImageLoadingState } from './types';

const logger = new Logger('Notifications.InAppMessaging');

const inAppMessageImageSizes: Record<InAppMessageLayout, number> = {
  BOTTOM_BANNER: BANNER_IMAGE_SCREEN_SIZE,
  MIDDLE_BANNER: BANNER_IMAGE_SCREEN_SIZE,
  TOP_BANNER: BANNER_IMAGE_SCREEN_SIZE,
  CAROUSEL: CAROUSEL_IMAGE_SCREEN_SIZE,
  FULL_SCREEN: FULL_SCREEN_IMAGE_SCREEN_SIZE,
  MODAL: MODAL_IMAGE_SCREEN_SIZE,
};

export const prefetchNetworkImage = async (
  url: string
): Promise<ImageLoadingState> => {
  try {
    const loaded = await Image.prefetch(url);
    if (loaded) {
      return 'loaded';
    }

    logger.error(`Image failed to load: ${url}`);
    return 'failed';
  } catch (e) {
    logger.error(`Image.prefetch failed: ${e}`);
    return 'failed';
  }
};

export const getLayoutImageDimensions = (
  imageHeight: number,
  imageWidth: number,
  layout: InAppMessageLayout
): ImageDimensions => {
  // determine aspect ratio for scaling rendered image
  const aspectRatio = imageWidth / imageHeight;
  const isSquare = aspectRatio === 1;
  const isPortrait = imageHeight > imageWidth;
  const isLandscape = imageWidth > imageHeight;

  // an image that has smaller dimensions than the max image dimension (e.g. 10px x 10px)
  // will be scaled up in size to match the size the message component expects.
  // While this could lead to pixelated images, it was ultimately a product decision,
  // ideally the message creator would follow the image guidelines in the pinpoint console
  const maxImageDimension = inAppMessageImageSizes[layout];

  let height: number;
  let width: number;

  // set square image dimensions
  if (isSquare) {
    height = maxImageDimension;
    width = maxImageDimension;
  }

  // set portrait image dimensions
  if (isPortrait) {
    height = maxImageDimension;
    width = maxImageDimension * aspectRatio;
  }

  // set landscape image dimensions
  if (isLandscape) {
    height = maxImageDimension / aspectRatio;
    width = maxImageDimension;
  }

  return { height, width };
};
