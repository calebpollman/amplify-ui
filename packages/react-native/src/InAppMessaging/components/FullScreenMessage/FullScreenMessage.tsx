/*
 * Copyright 2017-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

import React from 'react';
import { Image, Text, View } from 'react-native';

import icons from '../../../assets';
import TestIDs from '../../../AmplifyTestIDs';
import { Button, IconButton } from '../../ui';

import { ICON_BUTTON_HIT_SLOP, ICON_BUTTON_SIZE } from '../constants';
import { useMessageProps } from '../hooks';
import MessageWrapper from '../MessageWrapper';

import { getStyles } from './styles';
import { FullScreenMessageProps } from './types';

const { IN_APP_MESSAGING } = TestIDs;

export default function FullScreenMessage(props: FullScreenMessageProps) {
  const { body, header, image, onClose, primaryButton, secondaryButton } =
    props;
  const {
    hasButtons,
    hasPrimaryButton,
    hasRenderableImage,
    hasSecondaryButton,
    shouldRenderMessage,
    styles,
  } = useMessageProps(props, getStyles);

  if (!shouldRenderMessage) {
    return null;
  }

  return (
    <MessageWrapper style={styles.componentWrapper}>
      <View style={styles.container}>
        <View style={styles.contentContainer}>
          <IconButton
            color={styles.iconButton.iconColor}
            hitSlop={ICON_BUTTON_HIT_SLOP}
            onPress={onClose}
            size={ICON_BUTTON_SIZE}
            source={icons.close}
            style={styles.iconButton.container}
            testID={IN_APP_MESSAGING.CLOSE_BUTTON}
          />
          {hasRenderableImage && (
            <View style={styles.imageContainer}>
              <Image
                source={{ uri: image?.src }}
                style={styles.image}
                testID={IN_APP_MESSAGING.IMAGE}
              />
            </View>
          )}
          <View style={styles.textContainer}>
            {header?.content && (
              <Text style={styles.header} testID={IN_APP_MESSAGING.HEADER}>
                {header.content}
              </Text>
            )}
            {body?.content && (
              <Text style={styles.body} testID={IN_APP_MESSAGING.BODY}>
                {body.content}
              </Text>
            )}
          </View>
        </View>
        {hasButtons && (
          <View style={styles.buttonsContainer}>
            {hasSecondaryButton && (
              <Button
                onPress={secondaryButton.onPress}
                style={styles.secondaryButton.container}
                testID={IN_APP_MESSAGING.SECONDARY_BUTTON}
                textStyle={styles.secondaryButton.text}
              >
                {secondaryButton.title}
              </Button>
            )}
            {hasPrimaryButton && (
              <Button
                onPress={primaryButton.onPress}
                style={styles.primaryButton.container}
                testID={IN_APP_MESSAGING.PRIMARY_BUTTON}
                textStyle={styles.primaryButton.text}
              >
                {primaryButton.title}
              </Button>
            )}
          </View>
        )}
      </View>
    </MessageWrapper>
  );
}
