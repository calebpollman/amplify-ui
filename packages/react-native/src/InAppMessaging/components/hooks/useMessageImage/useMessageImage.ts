/*
 * Copyright 2017-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

import { useEffect, useRef, useState } from 'react';
import { Image } from 'react-native';

import { ConsoleLogger as Logger } from '@aws-amplify/core';
import {
  InAppMessageImage,
  InAppMessageLayout,
} from '@aws-amplify/notifications';

import { INITIAL_IMAGE_DIMENSIONS } from './constants';
import { ImageDimensions, ImagePrefetchStatus, UseMessageImage } from './types';
import { getLayoutImageDimensions, prefetchNetworkImage } from './utils';

const logger = new Logger('Notifications.InAppMessaging');

/**
 * Handles prefetching and dimension setting for message images
 *
 * @param {image} object - contains image source
 * @param layout - message layout
 * @returns {UseMessageImage} message image dimensions and rendering related booleans
 */

export default function useMessageImage(
  image: InAppMessageImage,
  layout: InAppMessageLayout
): UseMessageImage {
  const { src } = image ?? {};
  const shouldPrefetch = !!src;

  // set initial status to fetching if prefetch is required
  const [prefetchStatus, setPrefetchStatus] = useState<ImagePrefetchStatus>(
    shouldPrefetch ? ImagePrefetchStatus.FETCHING : null
  );
  const imageDimensions = useRef<ImageDimensions>(
    INITIAL_IMAGE_DIMENSIONS
  ).current;

  const isImageFetching = prefetchStatus === ImagePrefetchStatus.FETCHING;
  const hasRenderableImage = prefetchStatus === ImagePrefetchStatus.SUCCESS;

  useEffect(() => {
    if (!shouldPrefetch) {
      return;
    }

    prefetchNetworkImage(src).then((prefetchResult) => {
      if (prefetchResult === 'loaded') {
        // get image size once loaded
        Image.getSize(
          src,
          (imageWidth, imageHeight) => {
            const { height, width } = getLayoutImageDimensions(
              imageHeight,
              imageWidth,
              layout
            );
            imageDimensions.height = height;
            imageDimensions.width = width;

            setPrefetchStatus(ImagePrefetchStatus.SUCCESS);
          },
          (error) => {
            // handle size retrieval error
            logger.error(`Unable to retrieve size for image: ${error}`);
            setPrefetchStatus(ImagePrefetchStatus.FAILURE);
          }
        );
      } else {
        // handle prefetch failure
        setPrefetchStatus(ImagePrefetchStatus.FAILURE);
      }
    });
  }, [imageDimensions, layout, shouldPrefetch, src]);

  return { hasRenderableImage, imageDimensions, isImageFetching };
}
