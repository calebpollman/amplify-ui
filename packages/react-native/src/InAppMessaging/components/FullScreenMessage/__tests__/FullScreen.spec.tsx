/*
 * Copyright 2017-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

import React from 'react';
import TestRenderer from 'react-test-renderer';

import TestIDs from '../../../../AmplifyTestIDs';
import useMessageImage from '../../hooks/useMessageImage';
import { INITIAL_IMAGE_DIMENSIONS } from '../../hooks/useMessageImage/constants';

import FullScreenMessage from '../FullScreenMessage';

jest.mock('../../hooks/useMessageImage', () => ({
  __esModule: true,
  default: jest.fn(),
}));

const { IN_APP_MESSAGING } = TestIDs;

const mockUseMessageImage = useMessageImage as jest.Mock;
const onClose = jest.fn();
const onPress = jest.fn();

const baseProps = { layout: 'FULL_SCREEN' as const, onClose };

describe('FullScreenMessage', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders a message as expected without an image', () => {
    mockUseMessageImage.mockReturnValueOnce({
      hasRenderableImage: false,
      imageDimensions: INITIAL_IMAGE_DIMENSIONS,
      isImageFetching: false,
    });

    const renderer = TestRenderer.create(<FullScreenMessage {...baseProps} />);

    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('renders a message as expected with an image', () => {
    mockUseMessageImage.mockReturnValueOnce({
      hasRenderableImage: true,
      imageDimensions: { height: 100, width: 100 },
      isImageFetching: false,
    });

    const src = 'asset.png';
    const props = { ...baseProps, image: { src } };

    const renderer = TestRenderer.create(<FullScreenMessage {...props} />);

    const image = renderer.root.findByProps({ testID: IN_APP_MESSAGING.IMAGE });

    expect(image.props).toEqual(
      expect.objectContaining({ source: { uri: src } })
    );
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('returns null while an image is fetching', () => {
    mockUseMessageImage.mockReturnValueOnce({
      hasRenderableImage: false,
      imageDimensions: INITIAL_IMAGE_DIMENSIONS,
      isImageFetching: true,
    });

    const renderer = TestRenderer.create(<FullScreenMessage {...baseProps} />);

    expect(renderer.toJSON()).toBeNull();
  });

  it.each([
    [
      'header',
      IN_APP_MESSAGING.HEADER,
      { content: 'header content' },
      { children: 'header content' },
    ],
    [
      'body',
      IN_APP_MESSAGING.BODY,
      { content: 'body content' },
      { children: 'body content' },
    ],
    [
      'primaryButton',
      IN_APP_MESSAGING.PRIMARY_BUTTON,
      { onPress, title: 'primary button' },
      { children: 'primary button', onPress },
    ],
    [
      'secondaryButton',
      IN_APP_MESSAGING.SECONDARY_BUTTON,
      { onPress, title: 'secondary button' },
      { children: 'secondary button', onPress },
    ],
  ])('correctly handles a %s prop', (key, testID, testProps, expectedProps) => {
    mockUseMessageImage.mockReturnValueOnce({
      hasRenderableImage: false,
      imageDimensions: INITIAL_IMAGE_DIMENSIONS,
      isImageFetching: false,
    });

    const props = { ...baseProps, [key]: testProps };

    const renderer = TestRenderer.create(<FullScreenMessage {...props} />);
    const testElement = renderer.root.findByProps({ testID });

    expect(testElement.props).toEqual(expect.objectContaining(expectedProps));
  });

  it('calls onClose when the close button is pressed', () => {
    mockUseMessageImage.mockReturnValueOnce({
      hasRenderableImage: false,
      imageDimensions: INITIAL_IMAGE_DIMENSIONS,
      isImageFetching: false,
    });

    const renderer = TestRenderer.create(<FullScreenMessage {...baseProps} />);
    const closeButton = renderer.root.findByProps({
      testID: IN_APP_MESSAGING.CLOSE_BUTTON,
    });

    TestRenderer.act(() => {
      (closeButton.props as { onPress: () => void }).onPress();
    });

    expect(onClose).toHaveBeenCalledTimes(1);
  });
});
