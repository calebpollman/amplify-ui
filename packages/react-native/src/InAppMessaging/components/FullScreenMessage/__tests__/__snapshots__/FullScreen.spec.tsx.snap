// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FullScreenMessage renders a message as expected with an image 1`] = `
<Modal
  hardwareAccelerated={false}
  transparent={true}
  visible={true}
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Array [
        Object {
          "flex": 1,
        },
        Array [
          Object {
            "backgroundColor": "#fff",
          },
          Object {},
          Object {},
          Object {},
        ],
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flex": 1,
            "padding": 16,
          },
          Object {},
          Object {},
        ]
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          hitSlop={
            Object {
              "bottom": 10,
              "left": 10,
              "right": 10,
              "top": 10,
            }
          }
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignSelf": "flex-start",
                "marginBottom": 8,
                "marginLeft": "auto",
                "marginRight": 8,
              },
              undefined,
            ]
          }
          testID="aws-amplify__in-app-messaging--close-button"
        >
          <Image
            source={
              Object {
                "testUri": "../../../packages/react-native/src/icons/close.png",
              }
            }
            style={
              Object {
                "height": 20,
                "resizeMode": "contain",
                "tintColor": undefined,
                "width": 20,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "marginVertical": 12,
            }
          }
        >
          <Image
            source={
              Object {
                "uri": "asset.png",
              }
            }
            style={
              Array [
                Object {
                  "height": 100,
                  "width": 100,
                },
                undefined,
              ]
            }
            testID="aws-amplify__in-app-messaging--imager"
          />
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "marginHorizontal": 4,
              "marginVertical": 12,
              "paddingLeft": 8,
            }
          }
        />
      </View>
    </View>
  </RCTSafeAreaView>
</Modal>
`;

exports[`FullScreenMessage renders a message as expected without an image 1`] = `
<Modal
  hardwareAccelerated={false}
  transparent={true}
  visible={true}
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Array [
        Object {
          "flex": 1,
        },
        Array [
          Object {
            "backgroundColor": "#fff",
          },
          Object {},
          Object {},
          Object {},
        ],
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flex": 1,
            "padding": 16,
          },
          Object {},
          Object {},
        ]
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          hitSlop={
            Object {
              "bottom": 10,
              "left": 10,
              "right": 10,
              "top": 10,
            }
          }
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignSelf": "flex-start",
                "marginBottom": 8,
                "marginLeft": "auto",
                "marginRight": 8,
              },
              undefined,
            ]
          }
          testID="aws-amplify__in-app-messaging--close-button"
        >
          <Image
            source={
              Object {
                "testUri": "../../../packages/react-native/src/icons/close.png",
              }
            }
            style={
              Object {
                "height": 20,
                "resizeMode": "contain",
                "tintColor": undefined,
                "width": 20,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "marginHorizontal": 4,
              "marginVertical": 12,
              "paddingLeft": 8,
            }
          }
        />
      </View>
    </View>
  </RCTSafeAreaView>
</Modal>
`;
